crates/turborepo-lib/src/run/summary/spaces.rs

```diff
@@ -321,10 +326,33 @@ impl<'a> From<SpacesTaskInformation<'a>> for SpaceTaskSummary {
     }
 }
 
+fn trim_logs(logs: &[u8], limit: usize) -> String {
+    // Go JSON encoding automatically did a lossy conversion for us when
+    // encoding Golang strings into JSON.
+    let lossy_logs = String::from_utf8_lossy(logs);
+    if lossy_logs.as_bytes().len() <= limit {
+        lossy_logs.into_owned()
+    } else {
+        // We try to trim down the logs so that it is valid utf8
+        // We attempt to parse it at every byte starting from the limit until we get a
+        // valid utf8 which means we aren't cutting in the middle of a cluster.
+        for start_index in (lossy_logs.as_bytes().len() - limit)..lossy_logs.as_bytes().len() {
+            let log_bytes = &lossy_logs.as_bytes()[start_index..];
+            if let Ok(log_str) = std::str::from_utf8(log_bytes) {
```

Comment:
Is `from_utf8` O(n)? If so, this could be quite slow if the log file is quite large, say 2x the limit. Not really sure what the solution is, maybe a hard limit on log files that's 1.5x the spaces limit?

Reply:
It is O(n) on the path where the string is valid utf8, but it early exits when if finds an invalid character ([source](https://doc.rust-lang.org/src/core/str/validations.rs.html#155)). 

If we don't split a character, then we break out of the loop and only perform this O(n) operation once. 
If we do split a character, then `from_utf8` should exit after checking the remaining bytes in the char (up to 3) before it realizes it's invalid and we'll continue to do this until we are past the remaining bytes in the split character.